src/driver.rs

```diff
@@ -53,9 +54,24 @@ impl Driver {
         &self,
         py: Python<'_>,
         cluster: &magnum::Cluster,
+        upgrade: bool,
     ) -> PyResult<()> {
         py.allow_threads(|| {
             get_runtime().block_on(async {
+                // NOTE(mnaser): The updated cloud provider resource uses a different set of
+                //               labels and annotations (from the Helm chart) than the legacy
+                //               ones which was created by manifests. We need to clean up the
+                //               legacy resources otherwise it will generate a conflict during
+                //               the upgrade.
+                //
+                //               https://github.com/vexxhost/magnum-cluster-api/issues/580
+                if upgrade {
+                    debug!("Detecting cluster upgrade, ensuring that the legacy resource set is deleted");
+
+                    let client = cluster.client().await?;
+                    // TODO: make sure the "incorrect" resources are deleted if they exist
```

Comment:
This is the missing work here that I would say is more important, what needs to happen here is that we need to pull all the resources we used to install before via manifest, check if their matchLabels are the correct _new_ ones, if they're not, delete them.. because the next few lines will create the new cloud-provider secret.  we'll need to cover manila csi, cinder csi and cloud controller manager.